TECOPS-15052 AWS Customer Managed Keys (#1356)
snowplow/documentation: docs/pipeline/security/customer-managed-keys/index.md

---
title: "Customer managed keys"
sidebar_position: 1
---

This guide provides step-by-step instructions for setting up AWS KMS (Key Management Service) custom keys for your Snowplow pipeline. Custom KMS keys provide enhanced security by allowing you to maintain full control over encryption keys used by your data infrastructure.

:::note

This is a bolt-on security feature available for enterprise customers using BDP PMC (Private Managed Cloud) or BDP Cloud deployments.

:::

AWS Custom KMS Keys ensure that all data stored at-rest in AWS services (Kinesis, SQS, and S3) are encrypted with a key that you own and control. This provides:

- Enhanced security: full control over encryption keys
- Compliance: meet regulatory requirements for data encryption
- Audit control: ability to revoke access to encrypted data
- Key management: centralized control over data encryption

## Prerequisites

To use custom KMS keys, you will need:
- AWS account with appropriate permissions to create and manage KMS keys
- Access to your AWS Console in the same region as your Snowplow pipeline
- Snowplow pipeline account ID (provided by your Customer Success Manager)

> - Security team approval for cross-account key sharing

Reviewer: ```suggestion
- Your security team approval for cross-account key sharing
```

Author: Done

This feature has some limitations:

> - EMR shredder: if you're using the legacy EMR shredder process, it may not fully respect custom KMS keys for all output. This system is being deprecated.

Reviewer: Is this what is now known as RDB Transformer? (The non-streaming kind.) In that case, it’s not being deprecated for Redshift. So maybe we need to rephrase this to say that when loading to Redshift, keys might not be respected (what does that mean?).

Click **Create key** and configure with the following settings:

|       Setting       |                      Value                       |
| :-----------------: | :----------------------------------------------: |
|      Key type       |                    Symmetric                     |
|      Key usage      |               Encrypt and decrypt                |
| Key material origin |                       KMS                        |
|     Regionality     |                  Single region                   |
|        Alias        | `snowplow-pipeline-key` (or your preferred name) |

When setting key administrators:

- Key administrators: select the roles/users from your security team that need administrative access to the key
- Other AWS Accounts: add your Snowplow pipeline account ID here

:::note
The "Other AWS Accounts" step is essential. Without adding the Snowplow pipeline account ID, the encryption will not work.
:::

## Step 2: Key policy configuration

The key creation process will generate a policy similar to the one below. We strongly recommend allowing root access for the Snowplow pipeline account:

> ```json

Reviewer: Maybe this would be better in a collapsible block

Author: Added a collapsible block

  },
  "Action": [
    "kms:Encrypt",
    "kms:Decrypt",
    "kms:ReEncrypt*",
    "kms:GenerateDataKey*",
    "kms:DescribeKey",
    "kms:CreateGrant",
    "kms:ListGrants",
    "kms:RevokeGrant"
  ],
  "Resource": "*"
}
```

Using non-root policies requires coordination with Snowplow support to ensure all necessary roles are included. We strongly recommend the root access approach for simplicity.

## Step 3: Share key information with Snowplow

After creating your key, collect the following information:

- Key ARN: found in the key details page
  - Format: `arn:aws:kms:REGION:YOUR_ACCOUNT_ID:key/KEY_ID`
- Key ID: the unique identifier for your key
- Region: the AWS region where the key was created

Share this information with your Snowplow Customer Success Manager or through a support ticket:

- Key ARN
- Confirmation that cross-account access is configured
- Any specific requirements or restrictions

## Step 4: Snowplow configuration

Once you've shared the key information, Snowplow will configure the following pipeline components to use your custom KMS key.

All **pipeline S3 buckets** will be configured with your custom KMS key:
- Batch archive bucket
- Batch output bucket

> - Batch processing bucket

Reviewer: What are these? The ones for RDB Loader?

> - Kinesis S3 bad events bucket
> - Kinesis S3 enriched events bucket
> - Kinesis S3 raw events bucket

Reviewer: ```suggestion
- S3 bad events bucket
- S3 enriched events bucket
```
Let’s deleted the raw bucket since we are getting rid of it. Also “Kinesis S3” is confusing.

Author: Removed raw and removed the Kinesis prefix

Pipeline **Kinesis streams** will be encrypted with your key:
- Good events stream
- Bad events stream
- Enriched events stream

> - PII events stream

Reviewer: ```suggestion
```

Gone and not supported.

This guide provides step-by-step instructions for setting up AWS KMS (Key Management Service) custom keys for your Snowplow pipeline. Custom KMS keys provide enhanced security by allowing you to maintain full control over encryption keys used by your data infrastructure.

:::note

This is a bolt-on security feature available for enterprise customers using BDP PMC (Private Managed Cloud) or BDP Cloud deployments.

:::

AWS Custom KMS Keys ensure that all data stored at-rest in AWS services (Kinesis, SQS, and S3) are encrypted with a key that you own and control. This provides:

- Enhanced security: full control over encryption keys
- Compliance: meet regulatory requirements for data encryption
- Audit control: ability to revoke access to encrypted data
- Key management: centralized control over data encryption

## Prerequisites

To use custom KMS keys, you will need:
- AWS account with appropriate permissions to create and manage KMS keys
- Access to your AWS Console in the same region as your Snowplow pipeline
- Snowplow pipeline account ID (provided by your Customer Success Manager)
- Your security team approval for cross-account key sharing

This feature has some limitations:

> - EMR cluster: if you're using the legacy RDB Transformer process, it will need additional permissions adding (`kms:Decrypt`, `kms:GenerateDataKey*`) to enable access to customer S3 buckets with customer-managed KMS keys.

Reviewer: ```suggestion
- EMR cluster: if you're using the RDB Transformer component (only required for batch loading to Redshift and Databricks), it will need additional permissions adding (`kms:Decrypt`, `kms:GenerateDataKey*`) to enable access to customer S3 buckets with customer-managed KMS keys.
```

    ]
  },
  "Action": [
    "kms:Encrypt",
    "kms:Decrypt",
    "kms:ReEncrypt*",
    "kms:GenerateDataKey*",
    "kms:DescribeKey",
    "kms:CreateGrant",
    "kms:ListGrants",
    "kms:RevokeGrant"
  ],
  "Resource": "*"
}
```

</details>

Using non-root policies requires coordination with Snowplow Support to ensure all necessary roles are included. We strongly recommend the root access approach for simplicity.

## Step 3: Share key information with Snowplow

After creating your key, collect the following information:

- Key ARN: found in the key details page
  - Format: `arn:aws:kms:REGION:YOUR_ACCOUNT_ID:key/KEY_ID`
- Key ID: the unique identifier for your key
- Region: the AWS region where the key was created

Share this information with your Snowplow Customer Success Manager or through a support ticket:

- Key ARN
- Confirmation that cross-account access is configured
- Any specific requirements or restrictions

## Step 4: Snowplow configuration

Once you've shared the key information, Snowplow will configure the following pipeline components to use your custom KMS key.

All **pipeline S3 buckets** will be configured with your custom KMS key:
- Batch archive bucket

> - Batch output bucket (used in RDB Transformer/Loader process)
> - Batch processing bucket (used in RDB Transformer/Loader process)

Reviewer: ```suggestion
- Batch output bucket (used in RDB Transformer/Loader process for Redshift and Databricks)
- Batch processing bucket (used in RDB Transformer/Loader process for Redshift and Databricks)
```

- Drift application: Snowplow will apply necessary IAM policy updates to resolve access issues

You can verify the configuration by:
- Checking S3 bucket encryption settings in your AWS Console
- Verifying Kinesis stream encryption status
- Confirming SQS queue encryption (if applicable)
- Monitoring pipeline health through Snowplow dashboards

## Security best practices

For recommended settings:
- Enable key rotation: turn on automatic key rotation for enhanced security
- Monitor usage: use AWS CloudTrail to track key usage
- Regular audits: periodically review key policies and access patterns
- Documentation: maintain internal documentation of key management procedures

For key management:
- Backup procedures: ensure key policies are backed up and version controlled
- Access reviews: regular review of who has administrative access to keys
- Incident response: plan for key compromise or rotation scenarios

## Troubleshooting

If you encounter access denied errors:
- Verify the key policy includes the correct Snowplow pipeline account ID
- Ensure the key is in the same region as your pipeline
- Check that **Other AWS Accounts** was properly configured during key creation

> Pipeline failures after KMS implementation are expected temporarily while Snowplow applies the necessary IAM policy updates. Contact support if issues persist beyond the expected configuration window.

Reviewer 2: ```suggestion
Pipeline failures after KMS implementation are expected temporarily while Snowplow applies the necessary IAM policy updates. Contact Support if issues persist beyond the expected configuration window.
```

---
title: "Customer Managed Keys"
sidebar_position: 1
sidebar_label: Customer Managed Keys
---

## Overview

This guide provides step-by-step instructions for setting up AWS
KMS (Key Management Service) custom keys for your Snowplow pipeline.
Custom KMS keys provide enhanced security by allowing you to maintain
full control over encryption keys used by your data infrastructure.

:::note

> **Important**: This is a bolt-on security feature available for enterprise

Reviewer 2: why's there a line break here?

```suggestion
This is a bolt-on security feature available for enterprise
```
